File: web/src/api/network.ts
import { http } from '@/lib/http.ts';

export type DNSMode = 'manual' | 'dhcp';
export type EthernetMode = 'static' | 'dhcp';

// wake on lan
export function wol(mac: string) {
  const data = {
    mac
  };
  return http.post('/api/network/wol', data);
}

// get wake-on-lan macs history
export function getWolMacs() {
  return http.get('/api/network/wol/mac');
}

export function deleteWolMac(mac: string) {
  const data = {
    mac
  };
  return http.delete('/api/network/wol/mac', data);
}

// set Mac name
export function setWolMacName(mac: string, name: string) {
  return http.post('/api/network/wol/mac/name', { mac, name });
}

// get wifi information
export function getWiFi() {
  return http.get('/api/network/wifi');
}

// connect wifi without auth (only available in wifi configuration mode)
export function connectWifiNoAuth(ssid: string, password: string, apPassword?: string) {
  const data = {
    ssid,
    password
  };
  return http.post('/api/network/wifi', data, {
    headers: {
      'X-AP-Key': apPassword || ''
    }
  });
}

// verify ap login
export function verifyApLogin(apPassword: string) {
  return http.post(
    '/api/network/wifi/verify',
    {},
    {
      headers: {
        'X-AP-Key': apPassword || ''
      }
    }
  );
}

// connect wifi
export function connectWifi(ssid: string, password: string) {
  const data = {
    ssid,
    password
  };
  return http.post('/api/network/wifi/connect', data);
}

// disconnect wifi
export function disconnectWifi() {
  return http.post('/api/network/wifi/disconnect');
}

export function getDNS() {
  return http.get('/api/network/dns');
}

export function setDNS(mode: DNSMode, servers: string[]) {
  return http.post('/api/network/dns', { mode, servers });
}

export function getEthernet() {
  return http.get('/api/network/ethernet');
}

// Applies the addressing on trial. The device puts the previous settings back
// unless confirmEthernet arrives with this token before the window closes, so
// a wrong address costs a wait rather than a trip to the device.
export function setEthernet(
  mode: EthernetMode,
  address: string,
  prefix: number,
  gateway: string,
  trialSeconds?: number
) {
  return http.post('/api/network/ethernet', {
    mode,
    address,
    prefix,
    gateway,
    trialSeconds: trialSeconds ?? 0
  });
}

export function confirmEthernet(token: string) {
  return http.post('/api/network/ethernet/confirm', { token });
}

File: server/router/network.go
package router

import (
	"github.com/gin-gonic/gin"

	"NanoKVM-Server/authn"
	"NanoKVM-Server/middleware"
	"NanoKVM-Server/service/network"
)

func networkRouter(r *gin.Engine) {
	service := network.NewService()

	r.POST("/api/network/wifi", service.ConnectWifiNoAuth)    // connect Wi-Fi without auth (only available in ap mode)
	r.POST("/api/network/wifi/verify", service.VerifyApLogin) // verify ap login

	api := r.Group("/api").Use(middleware.CheckToken())

	api.POST("/network/wol", service.WakeOnLAN) // wake on lan
	api.GET("/network/wol/mac", service.GetMac) // get mac list

	admin := r.Group("/api").Use(
		middleware.CheckToken(),
		middleware.RequireRole(authn.RoleAdmin),
	)
	admin.DELETE("/network/wol/mac", service.DeleteMac)     // delete mac
	admin.POST("/network/wol/mac/name", service.SetMacName) // set mac name

	admin.GET("/network/wifi", service.GetWifi)                    // get Wi-Fi information
	admin.POST("/network/wifi/connect", service.ConnectWifi)       // connect Wi-Fi
	admin.POST("/network/wifi/disconnect", service.DisconnectWifi) // disconnect Wi-Fi
	admin.GET("/network/dns", service.GetDNS)                      // get DNS configuration
	admin.POST("/network/dns", service.SetDNS)                     // set DNS configuration

	// The address routes change the interface the caller is talking over, so
	// a change is applied on trial and reverts unless it is confirmed from
	// the new address.
	admin.GET("/network/ethernet", service.GetEthernet)              // get the wired address configuration
	admin.POST("/network/ethernet", service.SetEthernet)             // apply a wired address configuration on trial
	admin.POST("/network/ethernet/confirm", service.ConfirmEthernet) // keep the applied configuration
}
